docs/madr/decisions/039-transition-to-new-policies.md

```diff
@@ -0,0 +1,644 @@
+* as long as there is at least 1 policy of a new type (e.g. MeshTimeout), the policies of the corresponding old type (e.g. Timeout) are not applied
+  (i.e. if there is a MeshTimeout policy, the Timeout policies is not applied)
+
+The goal is to provide a way to transition from old policies to new policies. Having a smooth transition is going to enable 
+us to deprecate old policies in the future.
+
+Prerequisites for the transition:
+* both Global and Zone CPs are running the latest Kuma version
+
+## Decision Drivers
+
+- Predictability: when replacement policies are created (manually or automatically), user should be able to verify them 
+without applying for real
+- Short feedback loop: we'd like to deliver small portions of new policies more often, rather than big portions but rarely
+- Quick rollback: user should be able to rollback to old policies
+
+## Considered Options
+
+- Introduce a shadow mode for policies on Zone CP
+- Introduce a shadow mode for policies on Global CP
+  - Show potential changes for a specific DPP
+  - Show potential changes for all affected DPPs
+
+## Decision Outcome
+
+- Introduce a shadow mode for policies only on Zone CP
+
+## Shadow mode for policies
+
+Shadow mode is a way to test new policies without affecting the real traffic. When the policy is created with `kuma.io/effect: shadow` label
+
+```yaml
+type: MeshTimeout
+name: timeout-global
+mesh: default
+labels:
+  kuma.io/effect: shadow
+spec:
+  targetRef:
+    kind: Mesh
+  to: [...]
+```
+
+it doesn't apply on the real proxies. Instead, by using Inspect API we can inspect what config dump would look like if the shadow policies was applied.
+
+### Changes in the Inspect API
+
+Introduce a new non-physical endpoint `/_snapshot` that returns a config dump of a DPP without the actual `/config_dump` call to Envoy.
+If we want to be able to compute diff on the client side we can't use `xds` endpoint as its response slightly diverges from what CP has in the SnapshotCache.
+
+Add a query parameter `shadow=true`:
+
+* `GET /meshes/{mesh}/dataplanes/{name}/_snapshot` -> `GET /meshes/{mesh}/dataplanes/{name}/_snapshot?shadow=true`
+* `GET /meshes/{mesh}/dataplanes/{name}/_rules` -> `GET /meshes/{mesh}/dataplanes/{name}/_rules?shadow=true`
+
+when `shadow=true` is set, the response will take policies with `kuma.io/effect: shadow` label into account.
+
+For the debugging purposes it makes sense to represent the shadow snapshot as the original snapshot + jsonPatch that can be applied to it.
+Kuma GUI and kumactl also can benefit from such representation. To make server produce such response we can introduce a query parameter `include=diff`.
+Query param `include` contains a list of fields that should be included in the response. By default, only `xds` field is set by the API server. 
+From the validation point of view the request `_snapshot?shadow=true&include=diff` makes no sense and should be rejected.  
```

Comment:
It's the other way round no?

1. `_snapshot` or `_snapshot?shadow=false` : Returns the snaphost as is (no shadow policies)
2. `_snapshot?shadow=true` : Returns the snapshot as if shadow policies were enabled
3. `_snapshot?shadow=true&include=diff`: same as 2 but also returns the diff between this and what 1 would return
4. (not here but in the future) `_snapshot?proxy=true`: Returns the xds config from the proxy (equivalent to the current `xds` endpoint)
5. (not here but in the future) `_snapshot?proxy=true&include=diff`: 4 but also returns the diff between the proxy xds config and the snapshot of the CP (what 1 would return)
6. (not here but in the future) `_snapshot?proxy=true&shadow=true` ??? Is this even possible?

In light of 4,5,6 should we just have `_snapshot?mode=proxy|shadow` ? 
WDYT?

Reply:
Hmm isn't this what I say:

> 3. _snapshot?shadow=true&include=diff: same as 2 but also returns the diff between this and what 1 would return

Of course extra query_params doesn't change the rest of the response.

Reply:
I see, I got a wrong understanding of 3 initially, so `_snapshot?shadow=true&include=diff` returns
```yaml
xds: {...} # 'shadow' config
diff: [...] # jsonPatch that is applicable to 'xds' so you could get a 'current' config
```

But I guess if we call just `_snapshot?include=diff` then it returns:
```yaml
xds: {...} # 'current' config
diff: [...] # jsonPatch that is applicable to 'xds' so you could get a 'shadow' config
```

So in the first case `diff` is `shadow -> current` and in another `diff` is `current -> shadow`.

Reply:
Hmm to me `_snapshot?shadow=false&include=diff` would just return an empty diff as it compares current config with current config. I think it's just less suprising that the base of the diff remains unchanged (the CP snapshot as it's sent to the proxy)